src/journal/bsod.c

```diff
@@ -0,0 +1,249 @@
+                return log_error_errno(r, "Failed to read next journal entry: %m");
+        if (r == 0) {
+                log_debug("No emergency level entries in the journal");
+                return r;
+        }
+
+        r = sd_journal_get_data(j, "MESSAGE", &d, &l);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read journal message: %m");
+
+        message = memdup_suffix0((const char*)d + STRLEN("MESSAGE="), l - STRLEN("MESSAGE="));
+        if (!message)
+                return log_oom();
+
+        *ret = TAKE_PTR(message);
+
+        return 0;
+}
+
+static int find_next_free_vt(int fd, int *ret_free_vt, int *ret_original_vt) {
+        size_t i;
+        struct vt_stat terminal_status;
+
+        assert(fd);
+        assert(ret_free_vt);
+        assert(ret_original_vt);
+
+        if (ioctl(fd, VT_GETSTATE, &terminal_status) < 0)
+                return -errno;
+
+        for (i = 0; i < sizeof(terminal_status.v_state) * 8; i++)
+                if ((terminal_status.v_state & (1 << i)) == 0)
+                        break;
+
+        if (i >= sizeof(terminal_status.v_state) * 8)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTTY), "No free VT found: %m");
+
+        *ret_free_vt = i;
+        *ret_original_vt = terminal_status.v_active;
+
+        return 0;
+}
+
+static int display_emergency_message_fullscreen(char *message) {
+        int r, free_vt = 0, original_vt = 0;
+        unsigned qr_code_start_row = 1, qr_code_start_column = 1;
+        char tty[STRLEN("/dev/tty") + DECIMAL_STR_MAX(int) + 1];
+        _cleanup_close_ int fd = -EBADF;
+        _cleanup_fclose_ FILE *stream = NULL;
+        char read_character_buffer = '\0';
+        struct winsize w = {
+                .ws_col = 80,
+                .ws_row = 25,
+        };
+
+        assert(message);
+
+        fd = open_terminal("/dev/tty1", O_RDWR|O_NOCTTY|O_CLOEXEC);
+        if (fd < 0)
+                return log_error_errno(fd, "Failed to open tty1: %m");
+
+        r = find_next_free_vt(fd, &free_vt, &original_vt);
+        if (r < 0)
+                return log_error_errno(r, "Failed to find a free VT: %m");
+
+        xsprintf(tty, "/dev/tty%d", free_vt + 1);
+
+        fd = open_terminal(tty, O_RDWR|O_NOCTTY|O_CLOEXEC);
+        if (fd < 0)
+                return log_error_errno(fd, "Failed to open tty: %m");
+
+        if (ioctl(fd, TIOCGWINSZ, &w) < 0)
+                log_warning_errno(errno, "Failed to fetch tty size, ignoring: %m");
+
+        if (ioctl(fd, VT_ACTIVATE, free_vt + 1) < 0)
+                return log_error_errno(errno, "Failed to activate tty: %m");
+
+        r = loop_write(fd, ANSI_BACKGROUND_BLUE ANSI_HOME_CLEAR, strlen(ANSI_BACKGROUND_BLUE ANSI_HOME_CLEAR), /* do_poll = */ false);
+        if (r < 0)
+                log_warning_errno(r, "Failed to clear terminal, ignoring: %m");
+
+        qr_code_start_row = w.ws_row * 3U / 5U;
+        qr_code_start_column = w.ws_col * 3U / 4U;
```

Comment:
1. 60% height 75% width was gotten as the optimal position for my terminal window. lower numbers wouldn't be bottom-left enough, higher numbers would scatter the QR code drawing. I'd like to put some sort of fallback in place so that if 60%/75% doesn't work for a terminal, the fallback value is used and the QR code is saved.

Reply:
that's probably good enough for now